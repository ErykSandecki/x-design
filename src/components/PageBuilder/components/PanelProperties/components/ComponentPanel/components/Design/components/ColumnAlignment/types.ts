// types
import { AlignmentHorizontal, AlignmentVertical } from 'types';
import { TButtonGroup } from 'shared/UITools/components/ButtonGroup/types';

export type TAligmentHorizontalButtonGroup = Array<Pick<TButtonGroup, 'name'> & { key: AlignmentHorizontal }>;
export type TAligmentVerticalButtonGoup = Array<Pick<TButtonGroup, 'name'> & { key: AlignmentVertical }>;
