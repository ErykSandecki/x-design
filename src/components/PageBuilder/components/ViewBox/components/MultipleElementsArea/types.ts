// types
import { TElement } from 'types';

export type TCoordinatesData = {
  elementsCoordinates: Array<{
    coordinates: TRectCoordinates;
    id: TElement['id'];
  }>;
  outline: TRectCoordinates;
};
