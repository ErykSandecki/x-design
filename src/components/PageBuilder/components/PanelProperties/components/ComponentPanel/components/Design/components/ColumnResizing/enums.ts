export enum PopoverItem {
  auto = 1,
  fixed = 0,
  unit = 2,
}
