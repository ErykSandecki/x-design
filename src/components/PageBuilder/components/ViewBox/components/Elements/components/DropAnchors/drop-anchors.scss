@import '../../../../../../styles/z-indexes';

.DropAnchors {
  height: min-content;
  position: relative;
  width: min-content;

  &__prompt {
    background-color: #0b99ff;
    border-radius: 1px;
    box-sizing: border-box;
    position: absolute;
    z-index: $--xd-zindex-element-anchor;

    &--bottom {
      bottom: 0;
      height: 2px;
      left: 10px;
      width: calc(90% - 20px);
    }

    &--left {
      height: calc(90% - 20px);
      left: 0;
      top: 10px;
      width: 2px;
    }

    &--right {
      height: calc(90% - 20px);
      right: 0;
      top: 10px;
      width: 2px;
    }

    &--top {
      height: 2px;
      left: 10px;
      top: 0;
      width: calc(90% - 20px);
    }
  }

  &__anchor {
    position: absolute;
    z-index: $--xd-zindex-element-anchor;

    &--bottom {
      bottom: 0;
      height: 10px;
      left: 0;
      width: 100%;
    }

    &--left {
      height: 100%;
      left: 0;
      top: 0;
      width: 10px;
    }

    &--right {
      height: 100%;
      right: 0;
      top: 0;
      width: 10px;
    }

    &--top {
      height: 10px;
      left: 0;
      top: 0;
      width: 100%;
    }
  }
}
