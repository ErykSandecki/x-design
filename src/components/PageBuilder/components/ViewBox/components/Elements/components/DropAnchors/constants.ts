// types
import { DropAnchorsPosition } from './enums';

export const HORIZONTAL_ANCHORS = [
  DropAnchorsPosition.left,
  DropAnchorsPosition.right,
];
export const VERTICAL_ANCHORS = [
  DropAnchorsPosition.bottom,
  DropAnchorsPosition.top,
];
