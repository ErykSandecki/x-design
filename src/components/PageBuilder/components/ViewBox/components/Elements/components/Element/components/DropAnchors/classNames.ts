import { mapValues } from 'lodash';

// types
import { DropAnchorsPosition } from 'store/pageBuilder/enums';

export const className = 'DropAnchors';

export const classNames = {
  [className]: className,
  prompt: {
    name: `${className}__prompt`,
    modificators: {
      ...mapValues(DropAnchorsPosition, (position) => `${className}__prompt--${position}`),
    },
  },
  anchor: {
    name: `${className}__anchor`,
    modificators: {
      ...mapValues(DropAnchorsPosition, (position) => `${className}__anchor--${position}`),
    },
  },
};
