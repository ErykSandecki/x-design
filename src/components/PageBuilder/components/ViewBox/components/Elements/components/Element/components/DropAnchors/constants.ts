// types
import { DropAnchorsPosition } from 'store/pageBuilder/enums';

export const HORIZONTAL_ANCHORS = [
  DropAnchorsPosition.left,
  DropAnchorsPosition.right,
];

export const VERTICAL_ANCHORS = [
  DropAnchorsPosition.bottom,
  DropAnchorsPosition.top,
];

export const promptsData = (
  anchorPos: DropAnchorsPosition,
  displayNextPrompt: boolean,
  displayPrevPrompt: boolean,
  isFlowVertical: boolean,
  isGrid: boolean,
): Array<{ key: DropAnchorsPosition; visible: boolean }> => [
  {
    key: DropAnchorsPosition.top,
    visible:
      anchorPos === DropAnchorsPosition.top &&
      displayPrevPrompt &&
      (isGrid || isFlowVertical),
  },
  {
    key: DropAnchorsPosition.left,
    visible:
      anchorPos === DropAnchorsPosition.left &&
      displayPrevPrompt &&
      (isGrid || !isFlowVertical),
  },
  {
    key: DropAnchorsPosition.bottom,
    visible:
      anchorPos === DropAnchorsPosition.bottom &&
      displayNextPrompt &&
      (isGrid || isFlowVertical),
  },
  {
    key: DropAnchorsPosition.right,
    visible:
      anchorPos === DropAnchorsPosition.right &&
      displayNextPrompt &&
      (isGrid || !isFlowVertical),
  },
];
