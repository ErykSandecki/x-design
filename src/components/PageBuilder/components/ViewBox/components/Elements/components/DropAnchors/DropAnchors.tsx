import { FC, ReactNode } from 'react';
import { useSelector } from 'react-redux';

// components
import { E2EDataAttribute } from 'shared';

// hooks
import { useDropAnchorsEvents } from './hooks/useDropAnchorsEvents';
import { useTheme } from 'hooks';

// others
import { className, classNames } from './classNames';
import { HORIZONTAL_ANCHORS, VERTICAL_ANCHORS } from './constants';

// store
import {
  eventSelectorCreator,
  isDraggableSelectorCreator,
} from 'store/pageBuilder/selectors';

// styles
import styles from './drop-anchors.scss';

// types
import { DropAnchorsPosition } from './enums';
import { E2EAttribute, LayoutType, TElement } from 'types';
import { MouseMode } from 'types/enums/mouseMode';

// utils
import { enumToArray } from 'utils';

export type TDropAnchorsProps = {
  children: ReactNode;
  id: TElement['id'];
  index: number;
  mouseMode: MouseMode;
  layoutParent: TElement['layout'];
  parentId: TElement['parentId'];
};

const DropAnchors: FC<TDropAnchorsProps> = ({
  children,
  id,
  index,
  mouseMode,
  layoutParent,
  parentId,
}) => {
  const { type } = layoutParent;
  const { classNamesWithTheme, cx } = useTheme(classNames, styles);
  const { onMouseEnter, ...events } = useDropAnchorsEvents(index, mouseMode);
  const isDraggable = useSelector(isDraggableSelectorCreator(id));
  const possibleIndexPosition = useSelector(
    eventSelectorCreator('possibleIndexPosition'),
  ) as number;
  const possibleParent = useSelector(eventSelectorCreator('possibleParent'));
  const isFirst = index === 0;
  const isParent = possibleParent === parentId;
  const displayPrevPrompt =
    !isDraggable && isParent && isFirst && possibleIndexPosition === index;
  const displayNextPrompt =
    !isDraggable && isParent && possibleIndexPosition - 1 === index;
  const isVertical =
    type === LayoutType.default || type === LayoutType.vertical;
  const dropAnchorsPosition = isVertical
    ? VERTICAL_ANCHORS
    : HORIZONTAL_ANCHORS;

  return (
    <div className={cx(classNamesWithTheme[className])}>
      {displayPrevPrompt && (
        <div
          className={cx(
            classNamesWithTheme.prompt.name,
            [classNamesWithTheme.prompt.modificators.top, isVertical],
            [classNamesWithTheme.prompt.modificators.left, !isVertical],
          )}
        />
      )}
      {children}
      {displayNextPrompt && (
        <div
          className={cx(
            classNamesWithTheme.prompt.name,
            [classNamesWithTheme.prompt.modificators.bottom, isVertical],
            [classNamesWithTheme.prompt.modificators.right, !isVertical],
          )}
        />
      )}
      {enumToArray(DropAnchorsPosition)
        .filter((dropAnchor) =>
          dropAnchorsPosition.includes(dropAnchor as DropAnchorsPosition),
        )
        .map((position) => (
          <E2EDataAttribute
            key={position as keyof typeof DropAnchorsPosition}
            type={E2EAttribute.anchor}
            value={position as string}
          >
            <div
              className={cx(
                classNamesWithTheme.anchor.name,
                classNamesWithTheme.anchor.modificators[
                  position as keyof typeof DropAnchorsPosition
                ],
              )}
              onMouseEnter={() =>
                onMouseEnter(
                  DropAnchorsPosition[
                    position as keyof typeof DropAnchorsPosition
                  ],
                )
              }
              {...events}
            />
          </E2EDataAttribute>
        ))}
    </div>
  );
};

export default DropAnchors;
