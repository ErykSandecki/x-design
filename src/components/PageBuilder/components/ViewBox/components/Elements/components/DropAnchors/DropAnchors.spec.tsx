import { fireEvent, render } from '@testing-library/react';
import { Provider } from 'react-redux';

// components
import DropAnchors from './DropAnchors';

// mocks
import {
  elementDynamicDataMock,
  elementStaticDataMock,
  eventsMock,
  layoutMock,
  pageBuilderStateMock,
} from 'test/mocks/reducer/pageBuilderMock';
import { wholeStateMock } from 'test/mocks/reducer/wholeStateMock';
import 'test/mocks/sagas/allSagas';

// others
import { REDUCER_KEY as PAGE_BUILDER } from 'store/pageBuilder/actionsType';

// store
import { configureStore, store as storeToMock } from 'store/store';

// types
import { DropAnchorsPosition } from './enums';
import { E2EAttribute, LayoutType } from 'types';
import { MouseMode } from 'types/enums/mouseMode';

// utils
import { getByE2EAttribute } from 'test';

const mockCallBack = jest.fn();
const stateMock = {
  ...wholeStateMock,
  [PAGE_BUILDER]: {
    ...pageBuilderStateMock[PAGE_BUILDER],
    pages: {
      ...pageBuilderStateMock[PAGE_BUILDER].pages,
      ['0']: {
        ...pageBuilderStateMock[PAGE_BUILDER].pages['0'],
        elements: {
          dynamicData: { [elementDynamicDataMock.id]: elementDynamicDataMock },
          staticData: { [elementStaticDataMock.id]: elementStaticDataMock },
        },
      },
    },
  },
};

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockCallBack,
}));

describe('DropAnchors snapshots', () => {
  it('should render DropAnchors', () => {
    // mock
    const store = configureStore(stateMock);

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <DropAnchors
          id={elementDynamicDataMock.id}
          index={0}
          layoutParent={layoutMock}
          mouseMode={MouseMode.default}
          parentId={elementStaticDataMock.parentId}
        >
          element
        </DropAnchors>
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with prev prompt', () => {
    // mock
    const store = configureStore({
      ...stateMock,
      [PAGE_BUILDER]: {
        ...pageBuilderStateMock[PAGE_BUILDER],
        events: {
          ...eventsMock,
          draggableElements: [],
          possibleIndexPosition: 0,
          possibleParent: elementStaticDataMock.parentId,
        },
      },
    });

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <DropAnchors
          id={elementDynamicDataMock.id}
          index={0}
          layoutParent={layoutMock}
          mouseMode={MouseMode.default}
          parentId={elementStaticDataMock.parentId}
        >
          element
        </DropAnchors>
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with prev prompt', () => {
    // mock
    const store = configureStore({
      ...stateMock,
      [PAGE_BUILDER]: {
        ...pageBuilderStateMock[PAGE_BUILDER],
        events: {
          ...eventsMock,
          draggableElements: [],
          possibleIndexPosition: 0,
          possibleParent: elementStaticDataMock.parentId,
        },
      },
    });

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <DropAnchors
          id={elementDynamicDataMock.id}
          index={0}
          layoutParent={{ ...layoutMock, type: LayoutType.horizontal }}
          mouseMode={MouseMode.default}
          parentId={elementStaticDataMock.parentId}
        >
          element
        </DropAnchors>
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with next prompt', () => {
    // mock
    const store = configureStore({
      ...stateMock,
      [PAGE_BUILDER]: {
        ...pageBuilderStateMock[PAGE_BUILDER],
        events: {
          ...eventsMock,
          draggableElements: [],
          possibleIndexPosition: 1,
          possibleParent: elementStaticDataMock.parentId,
        },
      },
    });

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <DropAnchors
          id={elementDynamicDataMock.id}
          index={0}
          layoutParent={layoutMock}
          mouseMode={MouseMode.default}
          parentId={elementStaticDataMock.parentId}
        >
          element
        </DropAnchors>
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with next prompt', () => {
    // mock
    const store = configureStore({
      ...stateMock,
      [PAGE_BUILDER]: {
        ...pageBuilderStateMock[PAGE_BUILDER],
        events: {
          ...eventsMock,
          draggableElements: [],
          possibleIndexPosition: 1,
          possibleParent: elementStaticDataMock.parentId,
        },
      },
    });

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <DropAnchors
          id={elementDynamicDataMock.id}
          index={0}
          layoutParent={{ ...layoutMock, type: LayoutType.horizontal }}
          mouseMode={MouseMode.default}
          parentId={elementStaticDataMock.parentId}
        >
          element
        </DropAnchors>
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('DropAnchors behaviors', () => {
  it('should triger event mouse enter', () => {
    // mock
    const store = configureStore(stateMock);
    storeToMock.getState = () =>
      ({
        ...stateMock,
        [PAGE_BUILDER]: {
          ...stateMock[PAGE_BUILDER],
          events: {
            ...eventsMock,
            draggableElements: ['1'],
          },
        },
      }) as any;

    // before
    const { container } = render(
      <Provider store={store}>
        <DropAnchors
          id={elementDynamicDataMock.id}
          index={0}
          layoutParent={layoutMock}
          mouseMode={MouseMode.default}
          parentId={elementStaticDataMock.parentId}
        >
          element
        </DropAnchors>
      </Provider>,
    );

    // action
    fireEvent.mouseEnter(
      getByE2EAttribute(
        container,
        E2EAttribute.anchor,
        DropAnchorsPosition.top,
      ),
    );

    // result
    expect(mockCallBack.mock.calls[0][0].payload).toStrictEqual({
      possibleIndexPosition: 0,
    });
  });
});
