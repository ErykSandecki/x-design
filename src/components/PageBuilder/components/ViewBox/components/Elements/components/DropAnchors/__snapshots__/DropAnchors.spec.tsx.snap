// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropAnchors snapshots should render DropAnchors 1`] = `
<DocumentFragment>
  <div
    class="DropAnchors DropAnchors--light"
  >
    element
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--top DropAnchors__anchor--top--light"
      data-test-anchor="top"
    />
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--bottom DropAnchors__anchor--bottom--light"
      data-test-anchor="bottom"
    />
  </div>
</DocumentFragment>
`;

exports[`DropAnchors snapshots should render with next prompt 1`] = `
<DocumentFragment>
  <div
    class="DropAnchors DropAnchors--light"
  >
    element
    <div
      class="DropAnchors__prompt DropAnchors__prompt--light DropAnchors__prompt--bottom DropAnchors__prompt--bottom--light"
    />
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--top DropAnchors__anchor--top--light"
      data-test-anchor="top"
    />
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--bottom DropAnchors__anchor--bottom--light"
      data-test-anchor="bottom"
    />
  </div>
</DocumentFragment>
`;

exports[`DropAnchors snapshots should render with next prompt 2`] = `
<DocumentFragment>
  <div
    class="DropAnchors DropAnchors--light"
  >
    element
    <div
      class="DropAnchors__prompt DropAnchors__prompt--light DropAnchors__prompt--right DropAnchors__prompt--right--light"
    />
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--left DropAnchors__anchor--left--light"
      data-test-anchor="left"
    />
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--right DropAnchors__anchor--right--light"
      data-test-anchor="right"
    />
  </div>
</DocumentFragment>
`;

exports[`DropAnchors snapshots should render with prev prompt 1`] = `
<DocumentFragment>
  <div
    class="DropAnchors DropAnchors--light"
  >
    <div
      class="DropAnchors__prompt DropAnchors__prompt--light DropAnchors__prompt--top DropAnchors__prompt--top--light"
    />
    element
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--top DropAnchors__anchor--top--light"
      data-test-anchor="top"
    />
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--bottom DropAnchors__anchor--bottom--light"
      data-test-anchor="bottom"
    />
  </div>
</DocumentFragment>
`;

exports[`DropAnchors snapshots should render with prev prompt 2`] = `
<DocumentFragment>
  <div
    class="DropAnchors DropAnchors--light"
  >
    <div
      class="DropAnchors__prompt DropAnchors__prompt--light DropAnchors__prompt--left DropAnchors__prompt--left--light"
    />
    element
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--left DropAnchors__anchor--left--light"
      data-test-anchor="left"
    />
    <div
      class="DropAnchors__anchor DropAnchors__anchor--light DropAnchors__anchor--right DropAnchors__anchor--right--light"
      data-test-anchor="right"
    />
  </div>
</DocumentFragment>
`;
